File: packages/python/convai_analytics/resources/providers.py
"""Providers facade — provider/model comparison helpers."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

from ..measures import GroupBy, Segments, raw_value_percentile_measure
from ..types import BreakdownResponse

if TYPE_CHECKING:
    from ..client import ConvaiAnalytics


Component = Literal["llm", "tts", "asr", "neurosync"]

_COMPONENT_TO_SEGMENT: dict[Component, str] = {
    "llm": Segments.LLM_METRICS,
    "tts": Segments.TTS_METRICS,
    "asr": Segments.ASR_METRICS,
    "neurosync": Segments.NEUROSYNC_METRICS,
}


class ProvidersFacade:
    def __init__(self, client: ConvaiAnalytics) -> None:
        self._client = client

    def compare(
        self,
        *,
        component: Component = "llm",
        percentile: str = "p95",
        **filters: Any,
    ) -> BreakdownResponse:
        """"Compare LLM provider p95 latency over the last 7 days."

        Delegates to ``breakdown(group_by=provider|voiceProvider,
        segment=<component>Metrics)``.
        """
        return self._client.breakdown(
            measure=raw_value_percentile_measure(percentile),  # type: ignore[arg-type]
            # TTS uses `voiceProvider`; everything else uses `provider`.
            group_by=GroupBy.VOICE_PROVIDER if component == "tts" else GroupBy.PROVIDER,
            segment=_COMPONENT_TO_SEGMENT[component],
            **filters,
        )

File: packages/typescript/src/resources/providers.ts
/**
 * ProvidersFacade — provider/model comparison helpers.
 */

import type { ConvaiAnalytics } from "../client.js";
import {
  GROUP_BY,
  rawValuePercentileMeasure,
  SEGMENTS,
  type SegmentName,
} from "../measures.js";
import type {
  BreakdownResponse,
  CommonFilters,
  Percentile,
  TimeRange,
} from "../types.js";

export type Component = "llm" | "tts" | "asr" | "neurosync";

export interface ProviderCompareParams extends TimeRange, CommonFilters {
  /** Restrict to one component (e.g. 'llm', 'tts'). Default: 'llm'. */
  component?: Component;
  /** Default p95. */
  percentile?: Percentile;
}

const COMPONENT_TO_SEGMENT: Record<Component, SegmentName> = {
  llm: SEGMENTS.llmMetrics,
  tts: SEGMENTS.ttsMetrics,
  asr: SEGMENTS.asrMetrics,
  neurosync: SEGMENTS.neurosyncMetrics,
};

export class ProvidersFacade {
  readonly #client: ConvaiAnalytics;

  constructor(client: ConvaiAnalytics) {
    this.#client = client;
  }

  /**
   * "Compare LLM provider p95 latency over the last 7 days."
   * Delegates to `breakdown(groupBy=provider|voiceProvider, segment=<component>Metrics)`.
   */
  compare(params: ProviderCompareParams = {}): Promise<BreakdownResponse> {
    const { component = "llm", percentile = "p95", ...rest } = params;
    return this.#client.breakdown({
      measure: rawValuePercentileMeasure(percentile),
      // TTS uses `voiceProvider`; everything else uses `provider`.
      groupBy: component === "tts" ? GROUP_BY.voiceProvider : GROUP_BY.provider,
      segment: COMPONENT_TO_SEGMENT[component],
      ...rest,
    });
  }
}
